### PR TITLE
Start adding more complex flow tests

### PR DESCRIPTION
**Pull request by thieman**
@paetling @tleach This is how I was thinking about testing the interactions between the callbacks (thanks Alex for suggestion). Seem legit?

### CODE REVIEW

File: tests/fixtures.py [hunk 1]
```diff
@@ -1,10 +1,18 @@
...
-from collections import defaultdict
+from collections import defaultdict, namedtuple
+from uuid import uuid4
 
-from mock import Mock
+from mock import Mock, patch
 
 from kafka_rest.client import KafkaRESTClient
 
+class Callback(namedtuple('Callback', ['fn', 'args', 'kwargs'])):
```
Comment (paetling) [inline tests/fixtures.py]: what does wrapping a namedtuple in a class do?

Comment (thieman) [inline tests/fixtures.py]: Less janky way of throwing instance methods on it

File: tests/fixtures.py [hunk 2]
```diff
@@ -16,9 +24,55 @@ def with_basic_setup(cls):
         instance.registrar.emit = types.MethodType(_mock_emitter, instance)
         return instance
 
+    def install_mock_io_loop(self):
+        self.io_loop.stop()
+        self.io_loop = TestingIOLoop()
+
+        self.io_loop_patch = patch('kafka_rest.producer.IOLoop.current')
+        self.io_loop_mock = self.io_loop_patch.start()
+        self.io_loop_mock.return_value = self.io_loop
```
Comment (paetling) [inline tests/fixtures.py]: why do you need these patches?

Comment (thieman) [inline tests/fixtures.py]: tl;dr http://tornadokevinlee.readthedocs.org/en/latest/ioloop.html#tornado.ioloop.IOLoop.current

File: tests/fixtures.py [hunk 3]
```diff
@@ -16,9 +24,55 @@ def with_basic_setup(cls):
...
+        self.later.append(Later(handle, seconds, Callback(cb, args, kwargs)))
+        return handle
+
+    def remove_timeout(self, handle):
+        self.later = filter(lambda later: later.handle != handle, self.later)
+
+    @property
+    def next_callback(self):
+        if self.callbacks:
+            return self.callbacks[0]
```
Comment (paetling) [inline tests/fixtures.py]: do you want to pop here?

File: tests/fixtures.py [hunk 4]
```diff
@@ -16,9 +24,55 @@ def with_basic_setup(cls):
...
+            return self.later[0]
+
+    def run_next(self):
+        if self.callbacks:
+            self.callbacks[0].run()
+            del self.callbacks[0]
+
+    @property
+    def finished(self):
+        return not bool(self.callbacks or self.later)
```
Comment (paetling) [inline tests/fixtures.py]: You seem to never remove from callbacks?

File: tests/fixtures.py (hunk 3 above)
Comment (paetling) [inline tests/fixtures.py]: NVM.

File: tests/fixtures.py (hunk 4 above)
Comment (thieman) [inline tests/fixtures.py]: Yeah fixed that, pop in `run_next` now. Never remove from later, not sure how to handle that yet. Did it in that one test manually but should probably add methods for it. Not sure whether I need to bother doing shit in time order or not yet.

File: tests/flow_test.py [hunk 5]
```diff
@@ -0,0 +1,50 @@
...
+        self.client.teardown_mock_io_loop()
+
+    def test_produce_through_evaluate_no_flush(self):
+        self.client.produce('test_driver', self.test_value, self.test_schema)
+        self.assertEqual(self.io_loop.next_callback,
+                         Callback(self.producer.evaluate_queue,
+                                  ('test_driver', self.client.message_queues['test_driver']),
+                                  {}))
+
+        self.io_loop.run_next()
```
Comment (paetling) [inline tests/flow_test.py]: its seem strange that you checked a callback here, but ran a later

File: tests/fixtures.py [hunk 6]
```diff
@@ -16,9 +24,59 @@ def with_basic_setup(cls):
...
+            return self.laters[0]
+
+    def run_next(self):
+        if self.callbacks:
+            self.callbacks[0].run()
+            del self.callbacks[0]
+
+    def pop_later(self):
+        if self.laters:
+            del self.laters[0]
```
Comment (paetling) [inline tests/fixtures.py]: why do you never run the later functions?

**Pull request merged**